### PR TITLE
[SPARK-38089][CORE][TESTS] Show the root cause exception in `TestUtils.assertExceptionMsg`

### PR DESCRIPTION
**Pull request by xkrogen**
### What changes were proposed in this pull request?
Improve assertion failure message in `TestUtils.assertExceptionMsg` to print out the exception tree in which it was searching (upon failure).

### Why are the changes needed?
`TestUtils.assertExceptionMsg` is great, but when the assertion _doesn't_ match, it can be challenging to tell why, because the exception tree that was searched isn't printed. Only way I could find to fix it up was to run things in a debugger and check the exception tree. This makes it easier to tell what went wrong just from the assertion failure message.

### Does this PR introduce _any_ user-facing change?
No. Easier for devs to see why their test failed.

### How was this patch tested?
Used extensively while writing the tests for PR #34009. It was very useful!

For example, let's say I had a typo in the assertion for the test case `AvroSuite.SPARK-34133: Writing user provided schema respects case sensitivity for field matching`. The failure would look like:
```
java.lang.AssertionError: assertion failed: Exception tree doesn't contain the expected exception of type
org.apache.spark.sql.avro.IncompatibleSchemaException with message: Cannot find field "FOO" in Avro schema
	at scala.Predef$.assert(Predef.scala:223)
	at org.apache.spark.TestUtils$.assertExceptionMsg(TestUtils.scala:269)
	at org.apache.spark.sql.avro.AvroSuite.$anonfun$new$175(AvroSuite.scala:1409)
	...
```
So I know that I'm wrong, but I can't tell _why_. Now, after the change:
```
java.lang.AssertionError: assertion failed: Exception tree doesn't contain the expected exception of type
org.apache.spark.sql.avro.IncompatibleSchemaException with message: Cannot find field "FOO" in Avro schema
org.apache.spark.sql.avro.IncompatibleSchemaException: Cannot find field 'FOO' in Avro schema
	at org.apache.spark.sql.avro.AvroUtils$AvroSchemaHelper.$anonfun$validateNoExtraCatalystFields$1(AvroUtils.scala:265)
	at org.apache.spark.sql.avro.AvroUtils$AvroSchemaHelper.$anonfun$validateNoExtraCatalystFields$1$adapted(AvroUtils.scala:258)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.apache.spark.sql.avro.AvroUtils$AvroSchemaHelper.validateNoExtraCatalystFields(AvroUtils.scala:258)
	...

	at scala.Predef$.assert(Predef.scala:223)
	at org.apache.spark.TestUtils$.assertExceptionMsg(TestUtils.scala:266)
	at org.apache.spark.sql.avro.AvroSuite.$anonfun$new$175(AvroSuite.scala:1409)
	...
```
I can easily see that I used the wrong type of quotes. Much better!

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for spinning off this. Could you enable GitHub Action on your fork repository?

**Comment by dongjoon-hyun** [comment]
In addition, please include some example into the PR description. I have no idea how helpful it was to you.

**Comment by xkrogen** [comment]
Weird, GA is definitely enabled on my fork, I have used it many times. I pushed again a few minutes ago and it seems to be working again.
I also updated the description per your request.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs)
This will be okay if there is no outage in CIs due to the increase log size, @xkrogen .

**Comment by xkrogen** [comment]
Shouldn't be any change in log size unless there are test failures, since this will only print anything extra on assertion failure (which would cause a test failure). But of course let's wait for CI.

**Comment by xkrogen** [comment]
Looks like the test failures are some unrelated issues in the `yarn` tests:
```
[info] - run Spark in yarn-client mode with different configurations, ensuring redaction *** FAILED *** (3 minutes, 0 seconds)
[info]   The code passed to eventually never returned normally. Attempted 190 times over 3.0012781688500003 minutes. Last failure message: handle.getState().isFinal() was false. (BaseYarnClusterSuite.scala:204)
....
[info] - run Spark in yarn-cluster mode with different configurations, ensuring redaction *** FAILED *** (3 minutes, 0 seconds)
[info]   The code passed to eventually never returned normally. Attempted 190 times over 3.001207979333333 minutes. Last failure message: handle.getState().isFinal() was false. (BaseYarnClusterSuite.scala:204)
....
[info] *** Test still running after 2 minutes, 55 seconds: suite name: YarnClusterSuite, test name: yarn-cluster should respect conf overrides in SparkHadoopUtil (SPARK-16414, SPARK-23630). 
[info] - yarn-cluster should respect conf overrides in SparkHadoopUtil (SPARK-16414, SPARK-23630) *** FAILED *** (3 minutes, 0 seconds)
[info]   The code passed to eventually never returned normally. Attempted 190 times over 3.0010310271500003 minutes. Last failure message: handle.getState().isFinal() was false. (BaseYarnClusterSuite.scala:204)
....
[info] - SPARK-35672: run Spark in yarn-client mode with additional jar using URI scheme 'local' *** FAILED *** (3 minutes, 0 seconds)
[info]   The code passed to eventually never returned normally. Attempted 190 times over 3.0009631187333334 minutes. Last failure message: handle.getState().isFinal() was false. (BaseYarnClusterSuite.scala:204)
```
I'll re-trigger.

**Comment by dongjoon-hyun** [comment]
Thank you for re-triggering. I agree with you. I'll merge this when the CI passes, @xkrogen .

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.3.

**Comment by dongjoon-hyun** [comment]
BTW, for new feature and improvement JIRAs, `Affected Version` is `master` branch's version because Apache Spark doesn't allow backporting of new features and improvements.

**Comment by xkrogen** [comment]
Got it, thanks for the tip and for the review!

**Pull request closed without merging**
